File: Samples/Beyond.NET.Sample.Android/app/src/androidTest/java/com/example/beyondnetsampleandroid/SystemVersionTests.kt
package com.example.beyondnetsampleandroid

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.example.beyondnetsampleandroid.dn.*
import org.junit.Assert.*
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class SystemVersionTests {
    @Test
    fun testSystemVersionFromComponents() {

        val major = 1
        val minor = 2
        val build = 3
        val revision = 4

        val versionString = "$major.$minor.$build.$revision"

        val version = System_Version(major,
            minor,
            build,
            revision)

        val systemVersionType = System_Version.typeOf
        val versionFromComponentsType = version.getType()
        assertTrue(version.`is`(systemVersionType))
        assertTrue(systemVersionType == versionFromComponentsType)
        assertTrue(System_Object.equals(systemVersionType, versionFromComponentsType))

        val majorRet = version.major
        assertEquals(major, majorRet)

        val minorRet = version.minor
        assertEquals(minor, minorRet)

        val buildRet = version.build
        assertEquals(build, buildRet)

        val revisionRet = version.revision
        assertEquals(revision, revisionRet)

        assertEquals(versionString, version.dnToString().toKString())
    }
}

File: Samples/Beyond.NET.Sample.Swift/Tests/System/SystemVersionTests.swift
import XCTest
import BeyondDotNETSampleKit

final class SystemVersionTests: XCTestCase {
    override class func setUp() {
        Self.sharedSetUp()
    }
    
    override class func tearDown() {
        Self.sharedTearDown()
    }
    
    func testSystemVersionFromComponents() throws {
        let systemVersionType = System_Version.typeOf
        
        let major: Int32    = 1
        let minor: Int32    = 2
        let build: Int32    = 3
        let revision: Int32 = 4
        
        let versionString = "\(major).\(minor).\(build).\(revision)"
        
        let version = try System_Version(major,
                                         minor,
                                         build,
                                         revision)
        
        let versionFromComponentsType = try version.getType()
        
        guard systemVersionType == versionFromComponentsType else {
            XCTFail("System.Object.Equals should not throw and return true")
            
            return
        }
        
        let majorRet = try version.major
        XCTAssertEqual(major, majorRet)
        
        let minorRet = try version.minor
        XCTAssertEqual(minor, minorRet)
        
        let buildRet = try version.build
        XCTAssertEqual(build, buildRet)
        
        let revisionRet = try version.revision
        XCTAssertEqual(revision, revisionRet)
        
        let versionStringRet = try version.toString().string()
        XCTAssertEqual(versionString, versionStringRet)
    }
    
    func testSystemVersionFromString() throws {
        let major: Int32    = 123
        let minor: Int32    = 234
        let build: Int32    = 345
        let revision: Int32 = 456
        
        let versionString = "\(major).\(minor).\(build).\(revision)"
        let versionStringDN = versionString.dotNETString()
        
        let version = try System_Version(versionStringDN)
        
        let majorRet = try version.major
        XCTAssertEqual(major, majorRet)
        
        let minorRet = try version.minor
        XCTAssertEqual(minor, minorRet)
        
        let buildRet = try version.build
        XCTAssertEqual(build, buildRet)
        
        let revisionRet = try version.revision
        XCTAssertEqual(revision, revisionRet)
        
        let versionStringRet = try version.toString().string()
        XCTAssertEqual(versionString, versionStringRet)
    }
    
    func testSystemVersionParse() throws {
        let major: Int32    = 123
        let minor: Int32    = 234
        let build: Int32    = 345
        let revision: Int32 = 456
        
        let versionString = "\(major).\(minor).\(build).\(revision)"
        let versionStringDN = versionString.dotNETString()
        
        var version: System_Version?
        
        guard try System_Version.tryParse(versionStringDN,
                                          &version),
              let version else {
            XCTFail("System.Version.TryParse should not throw, return true and return an instance as out parameter")
            
            return
        }
        
        let majorRet = try version.major
        XCTAssertEqual(major, majorRet)
        
        let minorRet = try version.minor
        XCTAssertEqual(minor, minorRet)
        
        let buildRet = try version.build
        XCTAssertEqual(build, buildRet)
        
        let revisionRet = try version.revision
        XCTAssertEqual(revision, revisionRet)
        
        let versionStringRet = try version.toString().string()
        XCTAssertEqual(versionString, versionStringRet)
    }
}
